File: brickWall.py
# problem: https://leetcode.com/problems/brick-wall/
# Runtime: 180 ms, faster than 63.72% of Python3 online submissions for Brick Wall.
# Memory Usage: 19.2 MB, less than 37.29% of Python3 online submissions for Brick Wall.

from typing import List


class Solution:
    def leastBricks(self, wall: List[List[int]]) -> int:
        from collections import defaultdict
        rows = len(wall)
        edges = defaultdict(int)
        for row in wall:
            total = 0
            for i in range(len(row) - 1):
                total += row[i]
                edges[total] += 1
        maxVal = 0
        for val in edges.values():
            maxVal = max(maxVal, val)

        return rows - maxVal

File: brickWall.go
// problem: https://leetcode.com/problems/brick-wall/
// Runtime: 64 ms, faster than 60.87% of Go online submissions for Brick Wall.
// Memory Usage: 7.5 MB, less than 65.22% of Go online submissions for Brick Wall.

package algorithmss

func max(x int, y int) int {
	if x > y {
		return x
	}
	return y
}

func leastBricks(wall [][]int) int {
	num := 0
	quantities := map[int]int{}
	for _, row := range wall {
		for i := 0; i < len(row)-1; i++ {
			if i > 0 {
				row[i] += row[i-1]
			}
			quantities[row[i]] += 1
			num = max(quantities[row[i]], num)
		}
	}
	return len(wall) - num
}
